### PR TITLE
Release restrictions for BPFHostrouting when using the IPAMDelegatedPlugin

### PR DESCRIPTION
**Pull request by weiwei02**
The IPAMDelegatedPlugin plugin makes it possible not to write 
the code of the cloud vendor into the core code of Cilium. 
This mode is conducive to the scalability of Cilium. 
However:

1. DaemonConfig check requires that 
`LocalRouterIPv4`/`LocalRouterIPv6` must be set when 
IPAMDelegatedPlugin is used.
2. When `LocalRouterIPv6`/`LocalRouterIPv6` is set, 
the verifier requires that the `EnableEndpointRoutes=true` 
option must be set.
3. which is equivalent to prohibiting the use of BPFHostrouting by DelegateIPAM.

After turning off the `EnableEndpointRoutes` option, cloud providers 
actually have the opportunity to set their own routes. 

For example, in the cni plugin chain, the private cni plugin of the 
cloud provider manages the local routing, and cooperates with `cilium-cni `
to implement BPFHostRouting to accelerate the container network.

### CODE REVIEW

File: daemon/cmd/daemon_main.go
Comment (pippolo84) [inline daemon/cmd/daemon_main.go]: Nit: additional space
```suggestion
				log.Fatalf("Cannot specify %s or %s without %s.", option.LocalRouterIPv4, option.LocalRouterIPv6, option.EnableEndpointRoutes)
```

**Review by pippolo84 — APPROVED** [review]
LGTM apart a very small nit in the logging message.
Looped in @gandro to double check IPAM related changes.

**Review by gandro — CHANGES_REQUESTED** [review]
Thanks for the PR!

I'm generally not a fan of using a warning log message here. A warning should always be designed in a way where the user can fix something to make the warning go away.

Therefore, I think it would be better to add a flag instead, e.g. `--externally-managed-endpoint-routes` (there is probably a better name) for the user to indicate that the routes are managed by some external system.

@Weil0ng Could you please take a look as well, since you were the one who added the requirement to require EndpointRoutes.

**Review by gandro — COMMENTED** [review]
Thanks for the update, this looks good to me - assuming we actually need the old EndpointRoutes check (waiting on @Weil0ng to chime in if possible).

**Comment by gandro** [comment]
@weiwei02 Please do not request assigned reviewers. We still need a review from the `cli` code-owner team which I am not a part of.

**Comment by joestringer** [comment]
Thanks for your contribution.

How is network policy enforced in this mode? With `endpoint-routes=false`, Cilium expects that all pod traffic from the host will hit the cilium_host/cilium_net device pair, while with `endpoint-routes=true` we expect that the same traffic may be routed directly to the veth device for the Pod. If cloud provider routes direct the traffic directly towards Pods with endpoint routes disabled, then it could cause a network policy bypass upon ingress to application Pods.

**Comment by joestringer** [comment]
I would also be curious about the background here about why the restriction needs to exist in the first place. I think that's more important to understand and document the expectations, otherwise we may end up coming back in a few months because of a bug with the new flag and we won't understand how the flag fits into the overall routing model.

> For example, in the cni plugin chain, the private cni plugin of the
cloud provider manages the local routing, and cooperates with cilium-cni 
to implement BPFHostRouting to accelerate the container network.

Can you share any examples of how the local routing works in such a case?

When it comes to cooperating with cilium-cni to implement BPFHostRouting, what is missing? This will be supported with endpoint-routes mode in https://github.com/cilium/cilium/pull/22006 .

**Comment by joestringer** [comment]
> When LocalRouterIPv6/LocalRouterIPv6 is set,
the verifier requires that the EnableEndpointRoutes=true
option must be set.

I wonder if this is actually because if `endpoint-routes=false`, some other code in Cilium ignores the ``LocalRouterIPv4`/`LocalRouterIPv6` settings and just allocates an IP from the Pod IP range for use as the local IPv4/IPv6 routers. If that's the case, then I think that we should instead adjust that code to properly use the user-specified argument in this specific situation.

**Comment by weiwei02** [comment]
> Thanks for your contribution.
> 
> How is network policy enforced in this mode? With `endpoint-routes=false`, Cilium expects that all pod traffic from the host will hit the cilium_host/cilium_net device pair, while with `endpoint-routes=true` we expect that the same traffic may be routed directly to the veth device for the Pod. If cloud provider routes direct the traffic directly towards Pods with endpoint routes disabled, then it could cause a network policy bypass upon ingress to application Pods.

@joestringer Thank you for pointing out the problem. 
When using `endpoint-routes=false` then `to_ container` BPF section will not be attached to lxc_ xxx tc egress data path, which may cause the container's policy to be bypassed.

**Comment by weiwei02** [comment]
> I would also be curious about the background here about why the restriction needs to exist in the first place. I think that's more important to understand and document the expectations, otherwise we may end up coming back in a few months because of a bug with the new flag and we won't understand how the flag fits into the overall routing model.
> 
> > For example, in the cni plugin chain, the private cni plugin of the
> > cloud provider manages the local routing, and cooperates with cilium-cni
> > to implement BPFHostRouting to accelerate the container network.
> 
> Can you share any examples of how the local routing works in such a case?
> 
> When it comes to cooperating with cilium-cni to implement BPFHostRouting, what is missing? This will be supported with endpoint-routes mode in #22006 .

Thank you for helping to point out a better PR for using BPFHostRouting capability in IPAMDelegatedPlugin mode, #22006 can help me solve problems. I will close this pr

**Comment by Weil0ng** [comment]
> Thanks for the update, this looks good to me - assuming we actually need the old EndpointRoutes check (waiting on @Weil0ng to chime in if possible).

Sorry for late reply, as @joestringer said, the reason we limited the scope of `LocalRouterIP` is indeed because in other configurations it seems that the router IP is being used to SNAT the packets.

**Pull request closed without merging**
